lib/browse_everything/driver/sharepoint.rb

```diff
@@ -0,0 +1,232 @@
+      class << self
+        attr_accessor :authentication_klass
+
+        def default_authentication_klass
+          BrowseEverything::Auth::Sharepoint::Session
+        end
+      end
+
+      # Constructor
+      # @param config_values [Hash] configuration for the driver
+      def initialize(config_values)
+        self.class.authentication_klass ||= self.class.default_authentication_klass
+        super(config_values)
+      end
+
+      def icon
+        'cloud'
+      end
+
+      # Validates the configuration for the Sharepoint provider
+      def validate_config
+        raise InitializationError, 'Sharepoint driver requires a :client_id argument' unless config[:client_id]
+        raise InitializationError, 'Sharepoint driver requires a :client_secret argument' unless config[:client_secret]
+        raise InitializationError, 'Sharepoint driver requires a :tenant_id argument' unless config[:tenant_id]
+        raise InitializationError, 'Sharepoint driver requires a :redirect_uri argument' unless config[:redirect_uri]
+        raise InitializationError, 'Sharepoint driver requires a :scope argument' unless config[:scope]
+      end
+
+      # Retrieves the file entry objects for a given path to MS-graph drive resource
+      # @param [String] id of the file or folder
+      # @return [Array<BrowseEverything::FileEntry>]
+      def contents(id = '')
+        token_refresh if authorized?
+
+        folder = []
+        if id.empty?
+          folder << sites
+          folder << drives
+        else
+          folder << items_by_id(id)
+        end
+
+        values = []
+
+        folder.flatten.each do |f|
+          values << directory_entry(f)
+        end
+        @entries = values.compact
+
+        @sorter.call(@entries)
+      end
+
+      # @return [String]
+      # Authorization url that is used to request the initial access code from Sharepoint/Onedrive/365/etc
+      def auth_link(*_args)
+        Addressable::URI.parse("https://login.microsoftonline.com/#{config[:tenant_id]}/oauth2/v2.0/authorize?#{auth_query_string}")
+      end
+
+      # @return [Boolean]
+      def authorized?
+        @token.present?
+      end
+
+      def authorize!
+        return if @code.blank?
+        register_access_token(sharepoint_session.get_access_token(@code))
+        @code = nil
+        @token
+      end
+
+      def connect(params, _data, _url_options)
+        @code = params[:code]
+        authorize!
+      end
+
+      # @param [String] id of the file on MS graph drive
+      # @return [Array<String, Hash>]
+      def link_for(id)
+        file = items_by_id(id)
+        extras = {file_name: file['name'], file_size: file['size'].to_i}
+        [download_url(file), extras]
+      end
+
+      private
+
+      def auth_query_string
+        query = []
+        # keep_if deletes from and returns self, so dup config to not overwrite original
+        base = config.dup.keep_if { |k,v| ['client_id', 'scope', 'redirect_uri'].include?(k) }
```

Comment:
I think `slice` does the same thing as this line.
```suggestion
        base = config.slice('client_id', 'scope', 'redirect_uri')
```